File: ForgeSwift/ForgeCore/Sources/ForgeCore/Intelligence/AriaPersonalRead.swift
import Foundation

/// One coaching picture from the engines that already live in ForgeCore.
///
/// Dummy and local ARIA used to pick last-night hours and a population
/// readiness score. This fold is what they should actually coach from:
/// personal sleep depth, personal HRV, QoL, aging, swarm, and the ledger.
public struct AriaPersonalRead: Equatable, Sendable {
    public var sleepDepth: SleepDepthResult?
    public var sleepWeak: Bool
    public var hrvMs: Double?
    public var hrvBaselineMs: Double?
    public var hrvBelowPersonal: Bool
    public var keepLight: Bool
    public var companionLine: String?
    public var qolLine: String?
    public var agingLine: String?
    public var swarmLine: String?
    public var swarmProtect: Bool

    public init(
        sleepDepth: SleepDepthResult? = nil,
        sleepWeak: Bool = false,
        hrvMs: Double? = nil,
        hrvBaselineMs: Double? = nil,
        hrvBelowPersonal: Bool = false,
        keepLight: Bool = false,
        companionLine: String? = nil,
        qolLine: String? = nil,
        agingLine: String? = nil,
        swarmLine: String? = nil,
        swarmProtect: Bool = false
    ) {
        self.sleepDepth = sleepDepth
        self.sleepWeak = sleepWeak
        self.hrvMs = hrvMs
        self.hrvBaselineMs = hrvBaselineMs
        self.hrvBelowPersonal = hrvBelowPersonal
        self.keepLight = keepLight
        self.companionLine = companionLine
        self.qolLine = qolLine
        self.agingLine = agingLine
        self.swarmLine = swarmLine
        self.swarmProtect = swarmProtect
    }

    public var sleepBand: String {
        if sleepWeak { return "weak" }
        if let score = sleepDepth?.score {
            if score >= 80 { return "strong" }
            if score < 55 { return "weak" }
            return "ok"
        }
        return "unknown"
    }

    /// Grounding line ARIA can speak without dumping sensors.
    public var spokenGround: String? {
        if let companionLine, !companionLine.isEmpty { return companionLine }
        if let flag = sleepDepth?.headline { return flag }
        if hrvBelowPersonal { return "HRV is under your own baseline — not a population cutoff." }
        if swarmProtect, let swarmLine { return swarmLine }
        if let qolLine { return qolLine }
        return nil
    }

    public static func evaluate(
        nightHours: Double?,
        deepMinutes: Double?,
        remMinutes: Double?,
        awakeMinutes: Double?,
        hrvMs: Double?,
        readiness: Int,
        chronologicalAge: Double?,
        vo2Max: Double?,
        restingHR: Double?,
        swarm: AriaSwarmPicture? = nil,
        defaults: UserDefaults = .standard
    ) -> AriaPersonalRead {
        var sleepDepth: SleepDepthResult?
        if let hours = nightHours, hours > 0 {
            let metrics = SleepNightMetrics(
                totalHours: hours,
                deepMinutes: deepMinutes ?? 0,
                remMinutes: remMinutes ?? 0,
                efficiencyPercent: SleepNightMetrics.efficiencyPercent(
                    asleepHours: hours,
                    awakeMinutes: awakeMinutes ?? 0
                ),
                wakeConsistency: 80
            )
            let targets = SleepChronotypeTargets(
                targetHours: 8,
                deepGoalMinutes: 90,
                remGoalMinutes: 90
            )
            sleepDepth = SleepDepthScorer.score(
                metrics: metrics,
                targets: targets,
                baselines: SleepDepthBaselineStore.load(defaults: defaults)
            )
        }

        let hrvBaselines = BodyModelHRVBaselineStore.load(defaults: defaults)
        let hrvBaseline = hrvBaselines.baselineMs(.sdnn)
        let hrvBelow: Bool
        if let hrv = hrvMs, let base = hrvBaseline, base > 0 {
            hrvBelow = hrv < base * 0.92
        } else {
            hrvBelow = false
        }

        let sleepWeak = (nightHours ?? 9) < 6.5
            || (sleepDepth?.score ?? 100) < 55
            || !(sleepDepth?.unusualFlags.isEmpty ?? true)

        let qol = QualityOfLifeLivingStore.load(defaults: defaults)
        let qolLine: String? = qol == nil ? nil : QualityOfLifeLivingStore.coachingLine(defaults: defaults)
        let qolProtect = (qol?.overall ?? 100) < 55

        let aging = AgingSnapshot.evaluate(
            chronologicalAge: chronologicalAge,
            vo2Max: vo2Max,
            hrv: hrvMs,
            restingHR: restingHR,
            sleepHours: nightHours
        )
        let agingLine = aging.hasComparison ? aging.trainingHint : nil

        let swarmProtect = swarm.map { $0.stance.lowercased().contains("protect") || $0.stance.lowercased().contains("easy") } ?? false
        let swarmLine = swarm.flatMap { picture in
            let line = picture.headline.trimmingCharacters(in: .whitespacesAndNewlines)
            return line.isEmpty ? nil : line
        }

        let companion = AriaKnowledgeLedgerStore.load(defaults: defaults)
            .facts
            .sorted { $0.createdAt > $1.createdAt }
            .first(where: { $0.category == .weSpokeAbout || $0.category == .goals })
            .map(\.summary)

        let keepLight = sleepWeak || hrvBelow || qolProtect || swarmProtect || readiness < 55

        return AriaPersonalRead(
            sleepDepth: sleepDepth,
            sleepWeak: sleepWeak,
            hrvMs: hrvMs,
            hrvBaselineMs: hrvBaseline,
            hrvBelowPersonal: hrvBelow,
            keepLight: keepLight,
            companionLine: companion,
            qolLine: qolLine,
            agingLine: agingLine,
            swarmLine: swarmLine,
            swarmProtect: swarmProtect
        )
    }
}

File: backend/infra/lambda/aria_core/aria_personal_read.py
"""One coaching picture from the engines that already live in aria_core.

Python port of ForgeCore's ``AriaPersonalRead.swift``, store-free: callers
pass sleep-depth baselines, HRV baseline, QoL snapshot, swarm headline, and
companion line. UserDefaults stays on the phone.
"""

from __future__ import annotations

from dataclasses import dataclass

from . import sleep_depth_scorer as sds


@dataclass(frozen=True)
class PersonalRead:
    sleep_depth_score: int | None = None
    sleep_depth_headline: str | None = None
    sleep_weak: bool = False
    hrv_ms: float | None = None
    hrv_baseline_ms: float | None = None
    hrv_below_personal: bool = False
    keep_light: bool = False
    companion_line: str | None = None
    qol_line: str | None = None
    aging_line: str | None = None
    swarm_line: str | None = None
    swarm_protect: bool = False
    sleep_band: str = "unknown"

    @property
    def spoken_ground(self) -> str | None:
        if self.companion_line:
            return self.companion_line
        if self.sleep_depth_headline:
            return self.sleep_depth_headline
        if self.hrv_below_personal:
            return "HRV is under your own baseline — not a population cutoff."
        if self.swarm_protect and self.swarm_line:
            return self.swarm_line
        if self.qol_line:
            return self.qol_line
        return None

    def to_dict(self) -> dict:
        return {
            "sleepDepthScore": self.sleep_depth_score,
            "sleepDepthHeadline": self.sleep_depth_headline,
            "sleepWeak": self.sleep_weak,
            "hrvMs": self.hrv_ms,
            "hrvBaselineMs": self.hrv_baseline_ms,
            "hrvBelowPersonal": self.hrv_below_personal,
            "keepLight": self.keep_light,
            "companionLine": self.companion_line,
            "qolLine": self.qol_line,
            "agingLine": self.aging_line,
            "swarmLine": self.swarm_line,
            "swarmProtect": self.swarm_protect,
            "sleepBand": self.sleep_band,
            "spokenGround": self.spoken_ground,
        }


def evaluate(
    *,
    night_hours: float | None,
    deep_minutes: float | None = None,
    rem_minutes: float | None = None,
    awake_minutes: float | None = None,
    hrv_ms: float | None = None,
    hrv_baseline_ms: float | None = None,
    readiness: int = 0,
    qol_overall: int | None = None,
    qol_line: str | None = None,
    aging_line: str | None = None,
    swarm_line: str | None = None,
    swarm_stance: str | None = None,
    companion_line: str | None = None,
    sleep_baselines: sds.SleepDepthBaselines | None = None,
) -> PersonalRead:
    sleep_depth: sds.SleepDepthResult | None = None
    if night_hours is not None and night_hours > 0:
        metrics = sds.SleepNightMetrics(
            total_hours=night_hours,
            deep_minutes=deep_minutes or 0,
            rem_minutes=rem_minutes or 0,
            efficiency_percent=sds.efficiency_percent(night_hours, awake_minutes or 0),
            wake_consistency=80,
        )
        targets = sds.SleepChronotypeTargets(
            target_hours=8,
            deep_goal_minutes=90,
            rem_goal_minutes=90,
        )
        sleep_depth = sds.score(metrics, targets, sleep_baselines or sds.SleepDepthBaselines())

    hrv_below = False
    if hrv_ms is not None and hrv_baseline_ms is not None and hrv_baseline_ms > 0:
        hrv_below = hrv_ms < hrv_baseline_ms * 0.92

    sleep_weak = (night_hours if night_hours is not None else 9) < 6.5
    if sleep_depth is not None:
        sleep_weak = sleep_weak or sleep_depth.score < 55 or bool(sleep_depth.unusual_flags)

    qol_protect = (qol_overall if qol_overall is not None else 100) < 55
    swarm_protect = False
    if swarm_stance:
        lowered = swarm_stance.lower()
        swarm_protect = "protect" in lowered or "easy" in lowered

    keep_light = sleep_weak or hrv_below or qol_protect or swarm_protect or readiness < 55

    if sleep_weak:
        sleep_band = "weak"
    elif sleep_depth is not None:
        if sleep_depth.score >= 80:
            sleep_band = "strong"
        elif sleep_depth.score < 55:
            sleep_band = "weak"
        else:
            sleep_band = "ok"
    else:
        sleep_band = "unknown"

    return PersonalRead(
        sleep_depth_score=sleep_depth.score if sleep_depth else None,
        sleep_depth_headline=sleep_depth.headline if sleep_depth else None,
        sleep_weak=sleep_weak,
        hrv_ms=hrv_ms,
        hrv_baseline_ms=hrv_baseline_ms,
        hrv_below_personal=hrv_below,
        keep_light=keep_light,
        companion_line=companion_line,
        qol_line=None if qol_overall is None else qol_line,
        aging_line=aging_line,
        swarm_line=swarm_line,
        swarm_protect=swarm_protect,
        sleep_band=sleep_band,
    )
